### ISSUE TITLE
Updating volume of each audio track in a video on the fly as well as create a new file with final changes

### ISSUE THREAD

**Issue by Saket7042** [androidx/media#797]
I'm working on shorts/reels creation feature in that we have 3 different audio. Which are these

Original video sound
Voiceover which can be provided by user's mic
External audio clip which can be selected by user

I've to modify this on the fly(during playback) as well as create a new video file once user done with final changes. I'll look into this [ChannelMixingAudioProcessor](https://github.com/androidx/media/blob/release/libraries/common/src/main/java/androidx/media3/common/audio/ChannelMixingAudioProcessor.java).

So, to achieve this I've to pass volume float value to each Custom ChannelMixingAudioProcessor class for each audio tracks.

Just guessing there will be a few issues

I've to identify what is the type(Original, Voiceover, External audio) of audio track and based on the type what volume should be set.
As I'm trying to modify it on the fly and playing on Exoplayer so processing time may be long.

Please suggest some approach to implement this feature. Can someone can post an example by using ChannelMixingAudioProcessor that would be helpful.

**Comment by Samrobbo** [comment]
- `ChannelMixingAudioProcessor` is the recommended way to scale the volume of an audio stream, however the javadoc explains that the processor should not be interacted with or modified after it has been configured. Therefore you may need a custom AudioProcessor implementation (possibly wrapping or re-using some of the logic in ChannelMixingAudioProcessor).
- Audio processing is done on the players internal thread, meaning that any interactions that occur from the UI thread will need to handle the be done in a thread safe manner.

It's unclear whether you want to modify the volume of audio based on UI interactions, or whether you want to modify the volume based on the audio contents.

**Comment by Saket7042** [comment]
@Samrobbo I'm trying to build a feature in which we have a single video file with 3 audio tracks and on those audio tracks I've to manipulate the volume on the fly meaning User will be shown 3 sliders for each 3 audio tracks in UI with min 0 and max 100 volume and user can select the volume of each audio track. So once volume is selected by user then Video should render and play with that volume settings. Once user finalize the volume of each audio track then I've to generate a new video file.

**Comment by Samrobbo** [comment]
Just looking at one audio track and one volume slider: 
* You will need a custom AudioProcessor that has thread-safe communication with the UI.
* When the slider is changed, you'll have to inform the custom processor of the new volume.
* May need to trigger a flush of the AudioSink, depending on if you have the new volume applied instantly or as part of a pending value updated in flush.

This custom AudioProcessor will need to always be `isActive() == true`, and when there is no volume adjustment it should copy buffers to the output. There are some helper classes such as BaseAudioProcessor that may simplify things.

**Comment by Gustl22** [comment]
I have attempted a similar approach, see [this](https://github.com/google/ExoPlayer/issues/11329#issuecomment-1794529615). Unfortunately I won't have the time right now to test the `flush` on `AudioSink`. Plz keep the issue unlocked until validation 🙏🏻
